Device Manager logging improvements

swatisehgal opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?
/kind cleanup
<!--
Add one of the following kinds:
/kind bug
/kind cleanup
/kind documentation
/kind feature

Optionally add one or more of the following kinds if applicable:
/kind api-change
/kind deprecation
/kind failing-test
/kind flake
/kind regression
-->

#### What this PR does / why we need it:

Enhancing Device manager logs to ensure consistency across resource managers. This PR adds some missing metadata to logs, and makes use of klog.InfoS() and klog.ErrorS() as intended.

#### Which issue(s) this PR fixes:

<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Partially Fixes #123037
#### Special notes for your reviewer:

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs

```

swatisehgal committed:
node: device-mgr: Change ErrorS(nil, ...) to InfoS

Ensure consistency across resource managers and update
ErrorS(nil, ...) to InfoS. Similar changes have been
proposed in CPU Manager and Memory Manager.

Signed-off-by: Swati Sehgal <swsehgal@redhat.com>

swatisehgal committed:
node: device-mgr: Update klog.Infof(..., err) to klog.ErrorS(err,...)

Signed-off-by: Swati Sehgal <swsehgal@redhat.com>

swatisehgal committed:
node: device-mgr: Add metadata to logs

Ensure that if possible, we provide sufficient metadata
inclusing pod name and UID to allow filtering by pod name or its
UID.

Signed-off-by: Swati Sehgal <swsehgal@redhat.com>

swatisehgal committed:
node: device-mgr: Bump log level for expected skips

Signed-off-by: Swati Sehgal <swsehgal@redhat.com>

swatisehgal committed:
node: device-mgr: Keep log level consistent across all gRPC calls

Signed-off-by: Swati Sehgal <swsehgal@redhat.com>

ffromani reviewed (commented):
LGTM

I still have concerns about representing pod identity (namespace/name vs uid) and then we should stop for a sec before adding, because unfortunately adding is easier than removing.

I would call these "details". In general I'm very much in favor and overall happy with this much-needed improvement.

swatisehgal commented:
> LGTM
> 
> I still have concerns about representing pod identity (namespace/name vs uid) and then we should stop for a sec before adding, because unfortunately adding is easier than removing.
> 
> I would call these "details". In general I'm very much in favor and overall happy with this much-needed improvement.

Thanks for reviewing this so quickly!

I understand your concerns regarding how pod identity is represented. As you mentioned, adding is easy, but removing later could potentially impact end users. I’m definitely open to suggestions on improving this aspect.

That said, I do see value in having both representations. There are several parts of the code where we only have access to the `podUID` rather than the full pod object. This means that someone filtering logs by `name/namespace` alone might miss certain logs associated with the pod.

swatisehgal commented:
Pushed an update to resolve the lint failures. 

Pod identity representation is still an open for this PR and needs to be discussed. I will update the PR accordingly. This also applies to https://github.com/kubernetes/kubernetes/pull/129684 and https://github.com/kubernetes/kubernetes/pull/129685.

swatisehgal committed:
node: device-mgr: Add logs in the happy path

We have reasonable amount of logs when things go wrong.
While debugging, it can be useful to have logs to indicate that
things have gone as expected.

Signed-off-by: Swati Sehgal <swsehgal@redhat.com>

swatisehgal committed:
node: device-mgr: Update klog.Infof(..., err) to klog.ErrorS(err,...)

Signed-off-by: Swati Sehgal <swsehgal@redhat.com>

swatisehgal commented:
/test pull-kubernetes-unit

I made sure the unit tests are passing locally, let's retest

bart0sh commented on pkg/kubelet/cm/devicemanager/manager.go:
```diff
@@ -984,7 +987,7 @@ func (m *ManagerImpl) callPreStartContainerIfNeeded(podUID, contName, resource s
 	if eI.opts == nil || !eI.opts.PreStartRequired {
 		m.mutex.Unlock()
-		klog.V(4).InfoS("Plugin options indicate to skip PreStartContainer for resource", "resourceName", resource)
+		klog.V(5).InfoS("Plugin options indicate to skip PreStartContainer for resource", "podUID", podUID, "resourceName", resource, "containerName", contName)
```
the same here. Are there any guidelines regarding this?

swatisehgal replied:
For the same reasons as mentioned earlier: minimizing noise and maintaining consistency across resource managers. Please refer to the previous CPU Manager logging improvement for context: https://github.com/kubernetes/kubernetes/pull/120569/commits/2dc5ddd08a41a171134fd0922825ed03faa379ca.

Unfortunately, I haven't found guidelines/recommendations with this level of details in Kubernetes documentation. I know @pohly is quiet active in this area and might know?

I have been drafting some guidelines as I am working through the resource managers and am happy to contribute this for future reference but I am not sure where? :)

bart0sh replied:
I've seen some guidelines made by @pohly in the DRA code, but I don't remember where exactly. @pohly can you help?

pohly replied:
Guidelines for logging levels? https://github.com/kubernetes/community/blob/master/contributors/devel/sig-instrumentation/logging.md#what-method-to-use might be what you look for.

It's up to the code owners to decide which things are useful enough to be logged more often.

> I've seen some guidelines made by @pohly in the DRA code, but I don't remember where exactly. @pohly can you help?

Perhaps https://github.com/kubernetes/kubernetes/blob/814e2c89a80958bb0ba0ff186a41cda32cfe06da/staging/src/k8s.io/dynamic-resource-allocation/structured/allocator.go#L102-L106?

swatisehgal replied:
Perfect, Thank you!

bart0sh commented:
/triage accepted

swatisehgal commented on pkg/kubelet/cm/devicemanager/manager.go:
```diff
@@ -882,7 +885,7 @@ func (m *ManagerImpl) allocateContainerResources(pod *v1.Pod, container *v1.Cont
 		// TODO: refactor this part of code to just append a ContainerAllocationRequest
 		// in a passed in AllocateRequest pointer, and issues a single Allocate call per pod.
-		klog.V(3).InfoS("Making allocation request for device plugin", "devices", devs, "resourceName", resource)
+		klog.V(4).InfoS("Making allocation request for device plugin", "devices", devs, "resourceName", resource, "pod", klog.KObj(pod), "podUID", pod.UID, "containerName", container.Name)
```
We're adjusting the log level to reduce unnecessary noise while preserving valuable insights. This log captures an important flow, but the level of detail around allocation requests is primarily useful for debugging (enabled at log level 4).

As part of the broader effort to improve logging across all resource managers, consistency is a key focus. The goal is to enhance observability by ensuring that logs are valuable and capture meaningful state transitions for better debuggability without overwhelming the system with excessive noise. Striking this balance is critical.

swatisehgal committed:
node: device-mgr: Ensure consistent use of named arguments

Throughout the devicemanager codebase the named argument to represent
resource for logging pupose is `resourceName` as opposed to `resource`.
The latter can only be seen in topology_hints.go files. To ensure consistency
with the rest of the codebase and also because we want to adhere to the
recommendations in the Kubernetes documentation about named arguments:
https://github.com/kubernetes/community/blob/master/contributors/devel/sig-instrumentation/migration-to-structured-logging.md#name-arguments
we update the key from `resource` to `resourceName`.

Signed-off-by: Swati Sehgal <swsehgal@redhat.com>

swatisehgal committed:
node: device-mgr: Adhere to the message style guidelines

Ensure that the log messages adhere to the message style guildelines
as captured [here](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-instrumentation/logging.md#message-style-guidelines).

Signed-off-by: Swati Sehgal <swsehgal@redhat.com>

bart0sh commented:
/lgtm

/assign @ffromani @klueska 
for approval

ffromani commented on pkg/kubelet/cm/devicemanager/manager.go:
```diff
@@ -351,7 +351,7 @@ func (m *ManagerImpl) Start(activePods ActivePodsFunc, sourcesReady config.Sourc
 	err := m.readCheckpoint()
 	if err != nil {
-		klog.InfoS("Continue after failing to read checkpoint file. Device allocation info may NOT be up-to-date", "err", err)
+		klog.ErrorS(err, "Continue after failing to read checkpoint file. Device allocation info may NOT be up-to-date")
```
the nuance we lose here is the different log level in the logs, these will be now presented as ERRs which may be caught differently from alerts/monitoring tools. OTOH all this recoverable errors ARE worth an alert at least (we saw similar things in other resource managers).

Arguably, we should actually take the error path and do not swallow these errors, but this will change a long established behavior, so the fix can be worse than the issue.

TL;DR: I'm happier with this change than with the existing code.

swatisehgal replied:
I agree that changing the log level to `ERR` might impact monitoring systems differently, but given that these recoverable errors are still significant, I think they warrant better visibility. Other resource managers already follow a similar pattern, so this aligns with existing behavior elsewhere.

That said, I understand the concern about modifying long-standing behavior. I think we should capture this in the resource management logging guideline that I would like to publish soon. 

Let me know if you think there's a better way to handle this or if adding to the guideline addresses your concerns.

ffromani replied:
Clarification: the current change I think is beneficial for the reasons above. We may need a targeted revert of this very commit in the worst case, which is bearable.

The change of long-established behavior would be a followup hypotethical change on which we actually err out in the error paths on which currently we log ERR and continue. THAT would change a long established (debatable) behavior and THAT fix could likely be worse than the small "issue" this current change here is provoking.

So I think we should go along with this change with no further action until (if ever) further evidence comes out

ffromani commented on pkg/kubelet/cm/devicemanager/manager.go:
```diff
@@ -836,7 +836,7 @@ func (m *ManagerImpl) allocateContainerResources(pod *v1.Pod, container *v1.Cont
 		resource := string(k)
 		needed := int(v.Value())
-		klog.V(3).InfoS("Looking for needed resources", "needed", needed, "resourceName", resource)
+		klog.V(3).InfoS("Looking for needed resources", "resourceName", resource, "pod", klog.KObj(pod), "containerName", container.Name, "needed", needed)
```
worth recording in the commit message that we use podUID only when it is the only identifying info we have; if we have the full pod object, we prefer KObj and we defer to the contextual logging change the final decision about what to log, because always logging both can be redundant.

swatisehgal replied:
Yeah, good point. I am happy to update the commit message if you would like.

ffromani replied:
If for whatever reason you need to resubmit, please do. Otherwise, let's move on. This PR is LGTM+approved and in the merge pool, so let's go

ffromani reviewed (commented):
only minor comments from my side

ffromani commented:
/priority important-longterm
